### ISSUE TITLE
Stream.contains that returns index of element or -1

### ISSUE THREAD

**Issue by PromanSEW** [aNNiMON/Lightweight-Stream-API#112]
I created this function in my project:
```java
public static <T> int contains(List<T> list, Predicate<T> condition) {
	if (list.isEmpty()) return -1;
	for (int i=0; i < list.size(); i++) {
		if (condition.test(list.get(i))) return i;
	}
	return -1;
}
```
I use this function for example as
```java
protected int getPosition(String name) {
	return contains(getNames(), name::equalsIgnoreCase);
}
```
Where `getNames()` returns `List<String>`
What about adding something like this function as terminal operation for `Stream`?
Also `contains(T element)` will be useful
Maybe naming this feature as `indexOf` instead of `contains` will be better for describing its returning value

**Comment by aNNiMON** [comment]
I think it is helpful only for lists and arrays. If you use any intermediate operator before this function, the index will be different and useless.

```java
List<Integer> list = Arrays.asList(0, 1, 2, 3, 4, 5, 6, 7, 8);
int index = Stream.of(list)
        .filter(i -> i % 2 == 0)
        .contains(6);
// index = 3
```
In this example it is more reasonable to collect stream to list first, and then call `contains` on the result. Then the index can be used to access list items.

However, you can use `reduceIndexed` or custom operator.

```java
Predicate<String> p = "sam"::equalsIgnoreCase;
int index = Stream.of("Peter", "John", "Sam", "Kate")
        .reduceIndexed(-1, (i, acc, value) ->
                (acc >= 0) ? acc : (p.test(value) ? i : acc));
System.out.println(index); // 2


int index2 = Stream.of("Peter", "John", "Sam", "Kate")
        .custom(new ContainsOperator<>(p));

private static class ContainsOperator<T> implements Function<Stream<T>, Integer> {

    private final Predicate<? super T> predicate;

    public ContainsOperator(Predicate<? super T> predicate) {
        this.predicate = predicate;
    }

    @Override
    public Integer apply(Stream<T> stream) {
        IndexedIterator<? extends T> iterator = new IndexedIterator<T>(stream.iterator());
        while (iterator.hasNext()) {
            int index = iterator.getIndex();
            if (predicate.test(iterator.next())) {
                return index;
            }
        }
        return -1;
    }
}
```

**Comment by PromanSEW** [comment]
Thanks for the explanation. I forgot that intermediate operations can reduce `Stream` length and break indices
You can close this issue if it is not actual feature for implementation in your library

**Comment by aNNiMON** [comment]
You gave me an idea to implement `findFirstIndexed` operator. It will return `Optional<IntPair<T>>` with an element and its index. I can create two methods: `findFirstIndexed()` and `findIndexed(Predicate<T>)`.  

```java
Predicate<String> predicate = "sam"::equalsIgnoreCase;
int index = Stream.of("Peter", "John", "Sam", "Kate")
        .findIndexed(predicate)
        .ifPresent(p -> System.out.println(p.getFirst()));
```

**Comment by aNNiMON** [comment]
`findIndexed` implemented with #114
